### PR TITLE
Implement the authTokenCreate service

### PR DESCRIPTION
**Pull request by zenovich**
[authTokenCreate](https://docs.google.com/document/d/15Xxmatrs_7my5ioriikL6YgOU4I0sAVcq5ySKinc0Nw/edit?pli=1#heading=h.fitbll7av2ve)

### CODE REVIEW

File: app/api/auth/create_token.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		// usage of the old refresh token for that reason.
+		userMutex := make(chan bool)
+		defer close(userMutex)
+		userMutexInterface, loaded := userIDsInProgress.LoadOrStore(user.ID, userMutex)
+		// retry storing our mutex into the map
+		for ; loaded; userMutexInterface, loaded = userIDsInProgress.LoadOrStore(user.ID, userMutex) {
+			select { // like mutex.Lock(), but with cancel/deadline
+			case <-userMutexInterface.(chan bool): // it is much better than <-time.After(...)
+			case <-r.Context().Done():
+				return service.ErrUnexpected(r.Context().Err())
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/auth/create_token.go]: Yes, it is for request timeouts and connection breakage. Actually, for now, we don't have request timeouts and we don't handle broken connections, but I'll add logging here.

File: app/api/auth/create_token.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		service.MustNotBeError(srv.Store.InTransaction(func(store *database.DataStore) error {
+			sessionStore := store.Sessions()
+			// delete all the user's access tokens keeping the input token only
+			service.MustNotBeError(sessionStore.Delete("idUser = ? AND sAccessToken != ?",
+				user.ID, oldAccessToken).Error())
+			var err error
+			newToken, expiresIn, err = auth.CreateNewTempSession(sessionStore, user.ID)
+			return err
+		}))
+	} else {
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/auth/create_token.go]: Here we lock only token refreshing for only one user. So no need to reduce the locking duration here.

File: app/auth/temp_sessions.go [hunk 3]
```diff
@@ -32,5 +33,8 @@ func CreateNewTempSession(s *database.SessionStore, userID int64) (accessToken s
 		return
 	}
 
+	logging.Infof("Generated a session token %q expiring in %d seconds for a temporary user %d",
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/auth/temp_sessions.go]: We log access tokens on errors as well. Should I remove tokens from logs everywhere?

[the remark this answers is not in the document]
Comment (smadbe) [reply app/auth/temp_sessions.go]: Yes please :-). We could keep the first chars (and replace the others by "*") if we need to follow the token distribution, but it should not be needed.

File: app/api/auth/create_token.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply app/api/auth/create_token.go]: Yeah but it is still saner to keep the locked-code region to the minimum

File: app/api/auth/create_token.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/auth/create_token.go]: done

File: app/api/auth/create_token.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/auth/create_token.go]: fixed

File: app/auth/temp_sessions.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zenovich) [reply app/auth/temp_sessions.go]: fixed

File: app/api/auth/create_token.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		// We should not allow concurrency in this part because the login module generates not only
+		// a new access token, but also a new refresh token and revokes the old one. We want to prevent
+		// usage of the old refresh token for that reason.
+		userMutex := make(chan bool)
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/auth/create_token.go]: fixed

**Pull request merged**